File: LeetCode/0853.Car-Fleet/Car-Fleet.java
class Solution{
    public int target;
    class Pair{
        int x;
        int v;
        double t;
        public Pair(int xi, int vi) {
            x = xi; // int x (x)
            v = vi; // int y (x)
            t = (target - x + 0.0) / vi;        
        }
    }
    public int carFleet(int T, int[] X, int[] V) {// X: position V: speed T: target position
        target = T;
        int m = X.length;
        Pair[] cars = new Pair[m];
        for (int i = 0; i < m; i++) {
            cars[i] = new Pair(X[i], V[i]);
        }
        Arrays.sort(cars, (a,b)->(b.x - a.x)); //decending order

        int res = m;// car#
        int mCnt = 0;
        for (int i = 1; i < m; i++) {
            //double pre = cars[i - 1].t;
            //double cur = cars[i].t;
            if (Double.compare(cars[i].t, cars[i - 1].t) <= 0) {
            // [chialin] if (cars[i - 1].t > cars[i].t) {
            // System.out.printf(“%d %f”,i, cars[i].t);
                mCnt++;
                cars[i].t = cars[i - 1].t;
            }
        }
        res -= mCnt;
        return res;
    }
}

File: LeetCode/0853.Car-Fleet/Car-Fleet.py
class Solution(object):
    def carFleet(self, target, position, speed):
        """
        :type target: int
        :type position: List[int]
        :type speed: List[int]
        :rtype: int
        """
        # pos(t) = start + speed * t
        # pos(0, t) = 10 + 2t = 12, t = 1
        # pos(1, t) =  8 + 4t = 12, t = 1
        # pos(3, t) =  5 +  t = 12, t = 7
        # pos(4, t) =  3 + 3t = 12, t = 3
        # pos(2, t) =  0 +  t = 12, t = 12
        
        heap = []
        n = len(position)
        for i in range(n):
            heapq.heappush(heap, (-position[i], speed[i]))
        res = cur = 0
        while heap:
            p_, s = heapq.heappop(heap)
            p = -p_
            time = float(target - p) / s
            #print(time)
            if time <= cur: continue
            res += 1
            cur = time
            
        return res
                           
            
